Sprint 4 - Group support

johanxnilsson opened the pull request:
This adds groups support to the platform to isolate resources from users not within the same group.

sambles commented:
# Testing Feedback 


## Issues to investigate 

### 1.  Token refresh errors
i've seen this server error happen from both `admin` and `standard` user accounts intermittently 
```
HTTPError: 500 Server Error: Internal Server Error for url: http://api.oasis.local/refresh_token/
```
[4_crash_dump_refresh_tkn.txt](https://github.com/OasisLMF/OasisPlatform/files/7539642/4_crash_dump_refresh_tkn.txt)

### 2. Logging into the OasisUI with an admin account crashes the user interface 
This might be an OasisUI issue, however standard non-admin accounts seem to work until token refresh. (see 3.)

**Trace from the UI app**
```
oasisui API server: http://oasis-server:8000
In Login User:  oasis
Authenticated
.hideDivs step2 called
.hideDivs step3 called
.hideDivs step1 called
.defaultCreateProg called
.reloadtbl_portfoliosData called
Warning: Error in bind_rows_: Column `groups` can't be converted from character to list
  83: <Anonymous>
.reloadAnaData step2 called
```

### 3. OasisUI, standard user accounts timeout and crash the UI. 
This is probably connected to the token refresh issue, after about 5mins a standard account crashes the UI application.
Not much of note in the app trace. 
```
oasisui API server: http://oasis-server:8000
In Login User:  4_user
Authenticated
.hideDivs step3 called
.hideDivs step2 called
.hideDivs step1 called
.defaultCreateProg called
.reloadtbl_portfoliosData called
portfolio table refreshed
.reloadAnaData step2 called
.reloadAnaData landingpage called
analyses table refreshed
Navigate to Landing Page

(time out and crashed)

``` 

### 4. Open permission (no assignments) are not globally visible, items are hidden to standard users with groups assigned . 

**Example - `open_user`**  - Can view and access all objects with no groups assigned 
```
API Login:
Username: open_user
Password: 
Connecting to - http://api.oasis.local/

Available models:
+------+---------------+------------+--------------+
|   id | supplier_id   | model_id   | version_id   |
|------+---------------+------------+--------------|
|    1 | OasisLMF      | PiWind     | open         |
+------+---------------+------------+--------------+

Available portfolios:
+------+----------------+-----------------+-----------------+-------------------------+--------------------------+
|   id | name           | location_file   | accounts_file   | reinsurance_info_file   | reinsurance_scope_file   |
|------+----------------+-----------------+-----------------+-------------------------+--------------------------|
|    1 | open_portfolio | Yes             | -               | -                       | -                        |
+------+----------------+-----------------+-----------------+-------------------------+--------------------------+

Available analyses:
+------+---------------+---------+-------------+-----------------------------+--------------+---------------+----------------+
|   id | name          |   model |   portfolio | status                      | input_file   | output_file   | run_log_file   |
|------+---------------+---------+-------------+-----------------------------+--------------+---------------+----------------|
|    1 | open_analysis |       1 |           1 | INPUTS_GENERATION_STARTED   | -            | -             | -              |
  ... 
```

But all non-admin users with at least one group assigned are unable to see the **Open PiWind Model** or any object with no groups assigned.  
```
API Login:
Username: 1_user
Password: 
Connecting to - http://api.oasis.local/

Available models:
+------+---------------+------------+--------------+
|   id | supplier_id   | model_id   |   version_id |
|------+---------------+------------+--------------|
|    2 | OasisLMF      | PiWind     |            1 |
+------+---------------+------------+--------------+

Available portfolios:
+------+-------------+-----------------+-----------------+-------------------------+--------------------------+
|   id | name        | location_file   | accounts_file   | reinsurance_info_file   | reinsurance_scope_file   |
|------+-------------+-----------------+-----------------+-------------------------+--------------------------|
|    2 | 1_portfolio | Yes             | -               | -                       | -                        |
+------+-------------+-----------------+-----------------+-------------------------+--------------------------+

Available analyses:
+------+--------+---------+-------------+----------+--------------+---------------+----------------+
| id   | name   | model   | portfolio   | status   | input_file   | output_file   | run_log_file   |
|------+--------+---------+-------------+----------+--------------+---------------+----------------|
+------+--------+---------+-------------+----------+--------------+---------------+----------------+

```

### 5. Users with no groups assigned can see restricted analysis 
The `open_user` account in testing can view and download files from all analysis. However, the what should be restricted analysis are correctly read-only.  
```
Username: open_user
Password: 
Connecting to - http://api.oasis.local/

Available models:
+------+---------------+------------+--------------+
|   id | supplier_id   | model_id   | version_id   |
|------+---------------+------------+--------------|
|    1 | OasisLMF      | PiWind     | open         |
+------+---------------+------------+--------------+

Available portfolios:
+------+----------------+-----------------+-----------------+-------------------------+--------------------------+
|   id | name           | location_file   | accounts_file   | reinsurance_info_file   | reinsurance_scope_file   |
|------+----------------+-----------------+-----------------+-------------------------+--------------------------|
|    1 | open_portfolio | Yes             | -               | -                       | -                        |
+------+----------------+-----------------+-----------------+-------------------------+--------------------------+

Available analyses:
+------+---------------+---------+-------------+-----------------------------+--------------+---------------+----------------+
|   id | name          |   model |   portfolio | status                      | input_file   | output_file   | run_log_file   |
|------+---------------+---------+-------------+-----------------------------+--------------+---------------+----------------|
|    1 | open_analysis |       1 |           1 | INPUTS_GENERATION_STARTED   | -            | -             | -              |
|    2 | 2_analysis    |       3 |           5 | INPUTS_GENERATION_STARTED   | -            | -             | -              |
|    3 | 3_analysis    |       4 |           3 | INPUTS_GENERATION_CANCELLED | -            | -             | -              |
|    4 | 4_analysis    |       3 |           3 | INPUTS_GENERATION_STARTED   | -            | -             | -              |
|    5 | 5_analysis    |       5 |           4 | INPUTS_GENERATION_STARTED   | -            | -             | -              |
+------+---------------+---------+-------------+-----------------------------+--------------+---------------+----------------+
```

## Lower priority issues 

### 1. The Oasis-model helm charts can't create groups. 
The model auto-registration fails when assigning a group to a model that hasn't yet been populated by a user with that group authenticating first.  

### 2. Authenticating a user via the token request via the swagger endpoint `/access_token/` doesn't populate the django `auth_group` table. 
However It does work correct when a user logs in via the UI or directly authenticates via the API.

johanxnilsson commented:
## Issues to investigate
### 1. Token refresh errors

This happens when the refresh token has expired. I have replaced it with a proper 401 and error that says it has expired. The default access/refresh token timeout in keycloak is pretty short, about 5 minutes but can be increased.

### 2. Logging into the OasisUI with an admin account crashes the user interface

I could not reproduce this, but got a similar warning for all users:
```
Warning: Error in bind_rows_: Column `analysis_chunks` can't be converted from integer to character
  55: <Anonymous>
```
This greyed out the UI. I tried to add the missing fields in `tbl_analysesDataNames` but could not build the UI docker to test it. But maybe all we need to do is add those new fields to make it work?

### 3. OasisUI, standard user accounts timeout and crash the UI.

I think this is related to issue #1, but I could not verify it due to issue #2. I found this in the code:
```
api_fetch_response = function(meth, args, logMessage = print, ...) {
      response <- do.call(meth, eval(args, envir = sys.parent()))
      token_invalid <- status_code(response) == 401L
      # probably expired
      if (token_invalid) {
```
Which most likely didn't trigger the refresh since the api returned a 500 instead of 401. 

### 4. Open permission (no assignments) are not globally visible, items are hidden to standard users with groups assigned

This behavior is now changed to not treat "no groups" as a special group. All objects with no groups set are visible for all users, no matter group belongings.

### 5. Users with no groups assigned can see restricted analysis

Fixed this bug.

## Lower priority issues

### 1. The Oasis-model helm charts can't create groups.

This is now implemented. All admins can now provide non existing groups while creating new objects. This will create the groups and no longer reject them.

### 2. Authenticating a user via the token request via the swagger endpoint /access_token/ doesn't populate the Django auth_group table.

I had a quick look at this, and it is possible to add but needs a bit of work because /access_token/ is a public endpoint and not part of the keycloak security. I suggest we add this to a wish list.
